### ISSUE TITLE
second gtk dialog window does not receive keyboard focus

### ISSUE THREAD

**Issue by nomadicME** [awesomeWM/awesome#2220]
Output of `awesome --version`:

awesome v4.2 (Human after all)
 • Compiled against Lua 5.3.3 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.5
 • LGI version: 0.9.2

**How to reproduce the issue:**

Using the unmodified rc.lua provided in /etc/xdg/awesome/rc.lua, when a gtk dialog window is created it can be manipulated by the keyboard without any issues, and the window border is drawn correctly.  If this dialog is destroyed, then immediately following another dialog (identical to the first) is created and presented.  This dialog does not receive the keyboard focus nor is the theme border color/width applied (no border), it can be focused on using the "sloppy focus" of the mouse (which allows keyboard manipulation and border is drawn). What is interesting is, if you set titlebars_enabled = false to the empty rule in rules.rules (so it will be applied to every client) even sloppy focus does not allow the window to receive focus.

**Actual result:**

second dialog window does not receive focus.

**Expected result:**

it should receive focus

**Comment by psychon** [comment]
How are you creating dialogs that immediately open new dialogs when closed? Put differently: Which application do you want me to test and what exactly should I do with it?

**Comment by nomadicME** [comment]
I could not attach the python script that I hoped to submit with the
request.  I'll attach it to this email.  If it does not get through I'm
writing my own code editor gtk application in python. When I quit, I need a
prompt confirming that I do indeed want to quit.  Then it checks to see if
there are any unsaved buffers. If so, another dialog is presented asking if
I want to save.  This second dialog is the problem.  The attached script is
a very simple demonstration.  Click on the Hello World button which
presents the first dialog. Respond yes or no and the second one is
presented.

**Comment by psychon** [comment]
I'm sorry, but GitHub seems to ignore attachments to mails.

**Comment by Elv13** [comment]
You have to create a .zip and drag and drop it into the issue

**Comment by nomadicME** [comment]
OK, how about this:
[problem.gz](https://github.com/awesomeWM/awesome/files/1834760/problem.gz)

**Comment by nomadicME** [comment]
Just in case there is still trouble receiving the compressed file above, here is the contents of the python script:

#!/usr/bin/python

import gi
gi.require_version('Gtk', '3.0')

import signal
import sys
from gi.repository import Gtk as gtk
from gi.repository import Gdk as gdk
from gi.repository import GObject as gobject

class HelloWorld:

		def check_sigint_timer(self,timeout):
				gobject.timeout_add_seconds(timeout, self.check_sigint)

		def check_sigint(self):
				return True

		def hello(self, widget, data=None):
				print "Hello World"
# 				temp = dir(gtk)
# 				for t in temp:
# 					if 'ialog' in t:
# 						print t

				dialog = gtk.MessageDialog(parent=self.window, flags=gtk.DialogFlags.MODAL, type=gtk.MessageType.QUESTION, buttons=gtk.ButtonsType.YES_NO)	
				dialog.set_title("Question")
				dialog.set_position(gtk.WindowPosition.CENTER_ALWAYS)
				dialog.set_markup("Are you sure you want to quit ledit.py?")
				response = dialog.run()
				if response == gtk.ResponseType.YES:
					dialog.destroy()
				else:
					dialog.destroy()

				dialog = gtk.MessageDialog(parent=self.window, flags=gtk.DialogFlags.MODAL, type=gtk.MessageType.QUESTION, buttons=gtk.ButtonsType.YES_NO)	
				dialog.set_title("Question")
				dialog.set_position(gtk.WindowPosition.CENTER_ALWAYS)
				dialog.set_markup("Are you sure you want to quit ledit.py?")
				dialog.grab_focus()
				response = dialog.run()
				if response == gtk.ResponseType.YES:
					dialog.destroy()
				else:
					dialog.destroy()

		def quit_activated(self):
				gtk.main_quit()

		def delete_event_cb(self, widget, data=None):
				print "delete_event signal occurred"
				self.quit_activated()
				return True

		def destroy_cb(self, widget, data=None):
				print "destroy signal occurred"
				self.quit_activated()

		def signal_handler(self, signal, frame):
				print '\nYou pressed Ctrl+C!, exiting'
				gtk.main_quit()

		def on_key_press_cb(self,widget, event):
				keyname = gdk.keyval_name(event.keyval)
				if event.state & gdk.ModifierType.MOD1_MASK:
					pass
				elif event.state & gdk.ModifierType.CONTROL_MASK:
# # #  ( ctrl+q - quit )
					if keyname == 'q':	
						print 'entered ctrl+q'
						self.quit_activated()
# # #  ( ctrl+right - next tab right )
					elif keyname == 'Right':	
						print 'entered ctrl+right'

				else:
# # #  ( F2 - show slideshow of all images either on the card or the latest backup )
					if keyname == 'F2':
						self.change_bg()

		def __init__(self):

				signal.signal(signal.SIGINT, self.signal_handler)

				self.window = gtk.Window(gtk.WindowType.TOPLEVEL)
				self.window.connect("delete_event", self.delete_event_cb)
				self.window.connect("destroy", self.destroy_cb)
				self.window.connect('key_press_event', self.on_key_press_cb)
				self.window.set_border_width(10)
#				self.window.set_position(gtk.WindowPosition.CENTER_ALWAYS)
				self.window.maximize()

				self.button = gtk.Button("Hello World")
				self.button.connect("clicked", self.hello, None)
				self.window.add(self.button)
				self.button.show()

				self.window.show()

				self.check_sigint_timer(1)

		def main(self):
				gtk.main()

if __name__ == "__main__":
		hello = HelloWorld()
		hello.main()

**Comment by nomadicME** [comment]
Do I need to remove the "need info" label somehow?

**Comment by psychon** [comment]
Sorry for the long delay.

What I see so far:
- When "controlling things" with the mouse, the first dialog always gets the focus and so does the second one (since the mouse cursor already is inside of the second dialog when it opens, I guess)
- When using the keyboard to press buttons (space bar), i.e. when the mouse cursor is outside of the dialog when it opens, the first dialog gets focused and the second one does not. However, I can still control the second dialog with the keyboard since (I guess) GTK forwards the keyboard event internally.
- The same happens when I move the first dialog (so that the second dialog will not contain the pointer) and close it with the mouse
- Titlebars or not does not make a difference here

**Comment by nomadicME** [comment]
Sometimes I can control the second dialog when it does not receive the
focus, but most of the time I cannot.  So the fact that it does not receive
the focus seems to me to be a bug (especially for a wm designed to be used
with the keyboard) and should be addressed.

**Comment by psychon** [comment]
Finally figured this out. There was a complicated race going on that depended on the creation of the second dialog and the focus change to the main window (which in turn was caused by closing the first dialog) to be handled in the same main loop iteration.

Sorry that it took me so long.
